File: pyrect/util/_texture.f90
! -----------------------------------------------------------------------------
! echo.util._texture
! ==================
!
! Fortran routines for computing textures. A texture is defined as the standard
! deviation of a radar measurement within a 1-D or 2-D window centered around a
! radar gate.
!
! -----------------------------------------------------------------------------

! Necessary and/or potential future improvements to _texture submodule:
!
! * Properly handle contiguous sweep volumes, e.g., 360 deg PPI volumes.


subroutine compute_texture(data, sweep_start, sweep_end, ray_size, &
                           gate_size, rays_wrap_around, fill_value, &
                           debug, verbose, nr, ng, ns, std, sample_size)
! -----------------------------------------------------------------------------
! Compute texture of a radar measurement.
!
! Parameters
! ----------
! data : array, dim(nr, ng), float64
!   Radar data.
! sweep_start : array, dim(ns), int32
!   Sweep start ray indices indicating the start of a radar sweep.
! sweep_end : array, dim(ns), int32
!   Sweep end ray indices indicating the end of a radar sweep.
! ray_size : int32
!   Number of rays in the texture filter.
! gate_size : int32
!   Number of range gates in the texture filter.
! rays_wrap_around : logical
!   True if rays are contiguous in all sweeps.
! fill_value : float64
!   Value indicating missing or bad data in field.
! debug : logical
!   True to print debugging information, False to suppress.
! verbose : logical
!   True to print relevant information, False to suppress.
!
! Returns
! -------
! std : array, dim(nr, ng), float64
!   Texture of radar measurement.
! sample_size : array, dim(nr, ng), int32
!   Sample size in texture filter.
!
! -----------------------------------------------------------------------------
  implicit none

  integer(kind=4), intent(in)                     :: nr, ng, ns
  real(kind=8), intent(in), dimension(nr, ng)     :: data
  integer(kind=4), intent(in), dimension(ns)      :: sweep_start, sweep_end
  integer(kind=4), intent(in)                     :: ray_size, gate_size
  logical, intent(in)                             :: rays_wrap_around
  real(kind=8), intent(in)                        :: fill_value
  logical, intent(in)                             :: debug, verbose
  real(kind=8), intent(out), dimension(nr, ng)    :: std
  integer(kind=4), intent(out), dimension(nr, ng) :: sample_size

! Define local variables
  real(kind=8), parameter    :: atol=1.d-5
  integer(kind=4)            :: sweep, gate, ray, ray_start, ray_stop
  integer(kind=4)            :: r0, rf, g0, gf
  integer(kind=4)            :: n
  real(kind=8)               :: mu, var
  logical, dimension(nr, ng) :: valid_gate

! F2PY directives
! f2py integer(kind=4), optional, intent(in) :: nr, ng, ns
! f2py real(kind=8), intent(in)              :: data
! f2py integer(kind=4), intent(in)           :: sweep_start, sweep_end
! f2py integer(kind=4), intent(in)           :: ray_size, gate_size
! f2py logical, intent(in)                   :: rays_wrap_around
! f2py real(kind=8), intent(in)              :: fill_value
! f2py logical, intent(in)                   :: debug, verbose
! f2py real(kind=8), intent(out)             :: std
! f2py integer(kind=4), intent(out)          :: sample_size

  std = fill_value
  valid_gate = abs(data - fill_value) > atol

  do gate = 1, ng

    ! Gates along a ray are by definition not contiguous so no wrapping
    ! conditions have to be checked
    g0 = max(1, gate - gate_size / 2)
    gf = min(ng, gate + gate_size / 2)

    do sweep = 1, ns

    ray_start = sweep_start(sweep) + 1
    ray_stop = sweep_end(sweep) + 1

    do ray = ray_start, ray_stop

      r0 = ray - ray_size / 2
      rf = ray + ray_size / 2

      ! Check for condition where current ray is close to a sweep boundary
      if (r0 < ray_start .or. rf > ray_stop) then

        ! If sweep rays are contiguous then special care is needed
        if (rays_wrap_around) then
          stop
        else
         if (r0 < ray_start) then
            r0 = ray_start
          endif
          if (rf > ray_stop) then
            rf = ray_stop
          endif
        endif

      endif

      ! Compute sample size within texture filter
      n = count(valid_gate(r0:rf, g0:gf))
      sample_size(ray, gate) = n

      if (n > 1 .and. valid_gate(ray,gate)) then

      mu = sum(data(r0:rf, g0:gf), valid_gate(r0:rf, g0:gf)) / n
      var = sum((data(r0:rf, g0:gf) - mu)**2.d0, valid_gate(r0:rf, g0:gf)) / (n - 1)
      std(ray, gate) = sqrt(var)

      endif

    enddo
    enddo
  enddo

  return

end subroutine compute_texture

File: pyrect/util/texture.py
"""
pyrect.util.texture
===================

Submodule for computing textures of radar measurements. A texture is defined
as the standard deviation of a radar measurement within a 1-D or 2-D window
centered around a radar gate.

.. autosummary::
    :toctree: generated/

    add_textures
    _compute_texture

"""

import numpy as np
from scipy import ndimage

from pyart.config import get_field_name, get_metadata, get_fillvalue


def add_textures(radar, fields=None, gatefilter=None, size=(3, 3),
                 rays_wrap_around=False, debug=False, verbose=False):
    """
    Add texture fields to radar.

    Parameters
    ----------
    radar : pyart.core.Radar
        Radar containing specified fields.
    fields : sequence of strs, optional
        Radar fields. If None, all radar fields are used.
    gatefilter : pyart.filters.GateFilter, optional
        GateFilter specifying radar gates to exclude when computing texture.
    size : int or sequence of ints, optional
        The sizes of the texture filter are given for each axis as a sequence,
        or as a single number, in which case the size is equal for all axes.
        The default filter is 3 rays and 3 gates in size.

    Optional parameters
    -------------------
    rays_wrap_around : bool, optional
        True if rays are contiguous in all sweeps.
    debug : bool, optional
        True to print debugging information, False to suppress.
    verbose : bool, optional
        True to print relevant information, False to suppress.

    """

    # parse fields
    if fields is None:
        fields = radar.fields.keys()

    for field in fields:
        if verbose:
            print 'Computing texture field: {}'.format(field)

        _compute_texture(
            radar, field, gatefilter=gatefilter, size=size,
            rays_wrap_around=rays_wrap_around, debug=debug, verbose=verbose)

    return


def _compute_texture(radar, field, gatefilter=None, size=(3, 3),
                     rays_wrap_around=False, debug=False, verbose=False):
    """
    Compute texture field of radar measurement.

    Parameters
    ----------
    radar : pyart.core.Radar
        Radar containing specified field.
    field : str
        Radar field.
    gatefilter : pyart.filters.GateFilter, optional
        GateFilter specifying radar gates to exclude when computing texture.
    size : int or sequence of ints, optional
        The sizes of the texture filter are given for each axis as a sequence,
        or as a single number, in which case the size is equal for all axes.
        The default filter is 3 rays and 3 gates in size.

    Optional parameters
    -------------------
    rays_wrap_around : bool, optional
        True if rays are contiguous in all sweeps.
    debug : bool, optional
        True to print debugging information, False to suppress.
    verbose : bool, optional
        True to print relevant information, False to suppress.

    Notes
    -----

    """

    # parse image data
    image = radar.fields[field]['data']

    # Parse gate filter
    if gatefilter is not None:
        image = np.ma.masked_where(gatefilter.gate_excluded, image)

    if debug:
        N = np.ma.count(image)
        print 'Sample size of data: {}'.format(N)

    # parse image border parameter
    if rays_wrap_around:
        mode = 'wrap'
    else:
        mode = 'reflect'

    x = np.empty_like(image)  # store mean of image
    y = np.empty_like(image)  # store mean of image squared
    for slc in radar.iter_slice():
        ndimage.uniform_filter(
            image[slc], size=size, output=x[slc], mode=mode, origin=0)
        ndimage.uniform_filter(
            image[slc]**2.0, size=size, output=y[slc], mode=mode, origin=0)
    std = np.ma.sqrt(y - x**2.0)

    if debug:
        N = np.ma.count(std)
        print 'Sample size of texture field: {}'.format(N)

    # parse fill value
    fill_value = radar.fields[field].get('_FillValue', get_fillvalue())
    if np.ma.is_masked(std):
        std.set_fill_value(fill_value)

    std_dict = get_metadata(field)
    std_dict['data'] = std.astype(np.float32)
    std_dict['_FillValue'] = fill_value
    radar.add_field(
        '{}_texture'.format(field), std_dict, replace_existing=True)

    return
